### DISCUSSION TITLE
[SDK] <Can't deposit  to L2 with Zksync SDK>

### DISCUSSION THREAD

**Discussion by kadirchan** [zkSync-Community-Hub/zksync-developers#174] [SDKs] [answered]
### Environment

Testnet

### Acknowledgement

- [x] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

I want to write an app like zksync bridge frontend such as users connect their wallet and sign then deposit to l2 but when I look in SDK docs there is only examples with creating a wallet with PRIVATE_KEY and call deposit function. But I can't use users PRIVATE_KEY of course.

### Expected Behavior

User connect their metamask with using walletconnect modal then enter amount and sign deposit. Like bridge.zksync.io

### Code Example

const provider = new ethers.providers.Web3Provider(window.ethereum);
const signer = provider.getSigner();

const zkSigner = zksync.Signer.from(signer as any);

// ????? how to send deposit request to metamask

**Comment by MexicanAce** [comment]
Hi @kadirchan 👋 

Please take a look at using [Wagmi](https://wagmi.sh/) for connecting a user's wallet.

**Reply by kadirchan** [reply, marked as answer]
I solved it with migrate my app to react from angular thanks anyway

**Comment by patelkaren** [comment]
To enable users to interact with zkSync without exposing their private keys, you can utilize WalletConnect for wallet integration. Once the user connects their wallet via WalletConnect, you can obtain a signer object from the provider and use it to interact with zkSync functions. Here's a brief outline of the steps:

Use WalletConnect to connect the user's wallet.
Obtain the signer object from the connected provider.
Create a zkSync signer using zksync.Signer.from(signer).
Use the zkSync signer to call deposit functions, similar to the example you provided.
Here's a modified version of your code:  Connect wallet via WalletConnect
const provider = new ethers.providers.Web3Provider(window.ethereum);
const signer = await provider.getSigner();

 Create zkSync signer from the obtained signer object
const zkSigner = zksync.Signer.from(signer);

 Call deposit function using zkSync signer
 Example:
const token = "ETH";  Token symbol
const amount = ethers.utils.parseEther("1");  Amount to deposit
const deposit = await zkSigner.depositToSyncFromEthereum({
    depositTo: zkSigner.address(), // Deposit to the user's address
    token,
    amount,
});
console.log("Deposit transaction hash:", deposit.commit().hash);
  By following these steps, users can connect their wallets via WalletConnect and interact with zkSync without exposing their private keys.

**Comment by kaaron9** [comment]
To enable users to interact with zkSync using their connected wallets like MetaMask, you can utilize the Web3Provider and Signer from ethers.js along with the zkSync SDK. Here's a basic example of how you can handle deposit requests: const provider = new ethers.providers.Web3Provider(window.ethereum);
const signer = provider.getSigner();
const zkSyncProvider = await zksync.Provider.newWeb3Provider(provider);

 Get user's account address
const accountAddress = await signer.getAddress();

 Prompt user to deposit to L2
const depositAmount = ...; // Amount to deposit
const depositToken = "ETH"; // Token symbol (e.g., "ETH")

try {
     Initiate deposit transaction
    const deposit = await zkSyncProvider.getDepositContract(depositToken);
    const depositTx = await deposit.depositToSyncFromEthereum({
        depositTo: accountAddress,
        token: depositToken,
        amount: ethers.utils.parseEther(depositAmount.toString()), // Convert amount to wei
    

     Sign transaction with user's wallet
    const signedTx = await signer.signMessage(depositTx.ethMessage());
    
     Send signed transaction to zkSync
    const txHash = await zkSyncProvider.submitTx(signedTx);
    
    console.log("Deposit transaction submitted:", txHash);
 catch (error) {
    console.error("Error depositing to L2:", error);

 This code assumes that the user has already connected their MetaMask wallet using WalletConnect or a similar method. It prompts the user to specify the deposit amount and token symbol, then initiates a deposit transaction to L2 using zkSync's depositToSyncFromEthereum function. Finally, it signs the transaction with the user's connected wallet and submits it to zkSync.

Please note that this is a simplified example, and you may need to handle additional error cases, user input validation, and UI interactions in your application.

**Comment by golfdutoresste** [comment]
Assuming you have retrieved the amount and token to deposit
const amount = ...;  Amount to deposit
const token = ...;  Token to deposit (e.g., "ETH")

 Get the wallet instance
const wallet = new zksync.Wallet(zkSigner, zksync.ethers.getDefaultProvider("rinkeby")); // Use the appropriate network provider

Request deposit from the user's wallet
const deposit = await wallet.depositToSyncFromEthereum({
    depositTo: wallet.address(),  Deposit to the user's wallet address
    token,
    amount,


 Sign and submit the deposit request using Metamask
const ethereumTx = await deposit.syncTransfer.awaitReceipt();
console.log("Deposit transaction hash:", ethereumTx.txHash);
This code assumes you have already connected the user's wallet through WalletConnect and obtained a signer from it. You can then use that signer to create a zksync wallet instance. After that, you can use the depositToSyncFromEthereum method to initiate a deposit from Ethereum to Layer 2. Finally, you can await the receipt to get the transaction hash.

**Comment by hollinsmic** [comment]
The user wants to create an app similar to the zksync bridge frontend where users can connect their wallet, sign and deposit to L2. However, they are having trouble finding examples in the SDK docs that allow them to use the user's wallet without accessing their private key. They have provided a code example using ethers and zksync.Signer.from, but are unsure how to send a deposit request to Metamask.

**Reply by danijelTxFusion** [reply]
In the last few versions of `zksync-ethers` (v5 and v6), documented examples are provided for almost every function and method in the API. These examples can be easily accessed using an editor or IDE. Here are some of the examples:

- [Transfer](https://github.com/zksync-sdk/zksync-ethers/blob/main/src/signer.ts#L324),
- [Wthdraw](https://github.com/zksync-sdk/zksync-ethers/blob/main/src/signer.ts#L273),
- [Deposit](https://github.com/zksync-sdk/zksync-ethers/blob/main/src/signer.ts#L629).

**Comment by caojeremyv** [comment]
The user is trying to write an app similar to the zkSync bridge frontend, where users can connect their wallet and sign a deposit to L2. They are using the Go SDK and have only found examples that involve creating a wallet with a private key and calling the deposit function. The user is trying to figure out how to send a deposit request to Metamask without using the user's private key. The user has provided sample code that includes creating a Web3Provider and Signer, but is unsure of how to proceed with the deposit request.
